### DISCUSSION TITLE
Python 3.10 support for serverless functions?

### DISCUSSION THREAD

**Discussion by aorumbayev** [vercel/community#639] [Help] [answered]
Hello Vercel team and community members! 👋

Wanted to kick off this discussion to see if there is a planned roadmap for releasing support for python 3.10 in serverless function as its been around 2 years since support for 3.9 has been announced. 

Regards,
Al

**Comment by swarnava** [comment]
Hi, As soon as our upstream provider make it available, we will start supporting it. Unfortunately, we don't have any concrete ETA on this.

**Reply by aorumbayev** [reply]
@swarnava thanks for clarifying, looking forward to it! :)

**Reply by aorumbayev** [reply]
@swarnava any chance there has been any updates on this?

**Reply by swarnava** [reply]
It's still not available yet https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html. Thanks!

**Reply by aorumbayev** [reply]
@swarnava i assume still pending on updates from AWS side?

**Reply by swarnava** [reply]
Unfortunately, yes

https://github.com/aws/aws-lambda-base-images/issues/31
https://github.com/aws/aws-lambda-base-images/issues/62

**Reply by MrSlosky** [reply]
@swarnava, i can see PYthon 10 has been released, yet no changes made on Vercel? please uppgrade this.

**Reply by Waxime64** [reply]
python 3.10 is avalayble in AWS

**Comment by StigKorsnes** [comment]
Anytime soon perhaps. https://github.com/aws/aws-lambda-base-images/issues/31

**Comment by hsoiland** [comment]
https://github.com/aws/aws-lambda-base-images/issues/31#issuecomment-1448638312 fyi images have come through

**Comment by swarnava** [comment]
We are working with Build team to support this. ETA is still unknown. But we will update you as soon as it's available here and vercel.com/changelog. Thank you for your patience!

**Reply by aorumbayev** [reply]
Hello @swarnava, any updates on the discussion? 
Also what about 3.11 - is there a separate issue/thread to track when its going to be available on vercel?

**Reply by amyegan** [reply]
https://github.com/aws/aws-lambda-base-images/issues/31 and https://github.com/aws/aws-lambda-base-images/issues/62 are still open to add support for `3.10` and `3.11`. ⏳ It looks like `3.10` could be ready in the next couple of months based on the [latest official update](https://github.com/aws/aws-lambda-base-images/issues/31#issuecomment-1448638312). But there is [no official timeline](https://github.com/aws/aws-lambda-base-images/issues/62#issuecomment-1375773384) posted for `3.11` availability. We can continue the conversation here, but I expect it will be a while before we have news about `3.11`

**Reply by halvorhm** [reply]
Hi! It seems 3.10 was made available a couple of weeks ago. https://aws.amazon.com/blogs/compute/python-3-10-runtime-now-available-in-aws-lambda/

Looking forward to any news on serverless support!

**Reply by baseplate-admin** [reply]
Its august again. 

[AWS Lambda now supports python 3.11](https://aws.amazon.com/blogs/compute/python-3-11-runtime-now-available-in-aws-lambda/)

Do you folks have an ETA?

**Reply by dalist1** [reply]
Hi, any update on this?

**Reply by amyegan** [reply]
The team is working on support for `3.10` and `3.11` 🙇

**Reply by dalist1** [reply]
You guys started supporting Bun on its first week of its release and it's taking almost two years to upgrade the Python runtime 😅

**Reply by leerob** [reply]
We started supporting Bun as a _package manager_ right away. We do not yet support Bun as a first-party runtime, only experimental community support. https://github.com/vercel-community/bun

**Reply by tomnortonjr** [reply]
Good evening. Any updates on the Python runtime version upgrade? Thanks.

**Reply by amyegan** [reply]
Thank you all for waiting. I don't have an ETA for 3.10 or 3.11 support with automatic deployments, but we do have a way forward for anyone comfortable using Vercel CLI: https://github.com/vercel/vercel/pull/10648

**Reply by draew6** [reply]
@amyegan thanks for info

**Reply by adamstoffel** [reply]
@amyegan I reviewed the PR you linked to and I'm able to build python functions locally and deploy (using `vercel build` and `vercel deploy --prebuilt`).

However, I'm struggling because some libraries (in my case pydantic-core) are distributed as binaries so doing the local build causes binaries for my platform to download. I think the fact that the proper binaries for the Serverless Functions architecture are missing is causing an error like this (`Runtime.ImportModuleError: Unable to import module 'vc__handler__python': No module named 'pydantic_core._pydantic_core'`) when deployed to Vercel.

Any tips on getting the right platform binaries during the build?

**Reply by adamstoffel** [reply]
@amyegan  I'll also point out that the current implementation of ASGI in @vercel/python is incompatible with Python >=3.10. This means that frameworks like FastAPI won't work.

```
using Asynchronous Server Gateway Interface (ASGI)
[ERROR] TypeError: As of 3.10, the *loop* parameter was removed from Queue() since it is no longer necessary
Traceback (most recent call last):
  File "/var/task/vc__handler__python.py", line 305, in vc_handler
    response = asgi_cycle(__vc_module.app, body)
  File "/var/task/vc__handler__python.py", line 202, in __call__
    self.app_queue = asyncio.Queue(loop=loop)
  File "/var/lang/lib/python3.10/asyncio/queues.py", line 34, in __init__
    super().__init__(loop=loop)
  File "/var/lang/lib/python3.10/asyncio/mixins.py", line 17, in __init__
    raise TypeError(
```

**Comment by malazaysc** [comment]
Hello! any updates on 3.10 and 3.11?

**Comment by treehousekingcomic** [comment]
Oh its June again
Any updates on 3.10?

**Comment by XChikuX** [comment]
Is it really that hard to change runtimes? 
It's generally a trivial matter. Why is it taking so long, especially if the underlying functions are already provided by AWS ?

At this point you should skip 3.10 and 3.11 altogether and directly deploy 3.12. It's vastly superior due to the removal of 'GIL'

**Comment by msoedov** [comment]
Any update on 3.10? 3.12 is already there

**Comment by zsiegel92** [comment]
[Python 3.11 is supported by AWS Lambda](https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html).

Considering the massive popularity of Pydantic and FastAPI, which have [vastly different type annotation APIs/behavior between Python 3.9 and 3.10](https://fastapi.tiangolo.com/yo/python-types/), I think many people would greatly appreciate adoption of newer Lambda runtimes by Vercel.

I have no knowledge of how Vercel works this out, but I do wonder whether it would be possible to specify a *Lambda runtime* ([AWS Lambda docs](https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html)) in `vercel.json` rather than or in addition to a *`python_version`* in the `Pipfile` ([Vercel docs](https://vercel.com/docs/functions/serverless-functions/runtimes/python)).

Alternatively, rather than handling declarative config like this, perhaps "[custom install commands for your API](https://vercel.com/docs/deployments/configure-a-build#custom-install-command-for-your-api)" would be the right entrypoint:

> The Install Command defined in the Project Settings will be used for front-end frameworks that support Serverless Functions for APIs.
>
> If you're using [Serverless Functions](https://vercel.com/docs/functions/serverless-functions) defined in the natively supported api directory, a different Install Command will be used depending on the language of the Serverless Function. You cannot customize this Install Command.

Maybe Vercel can provide or expose some APIs to let users customize their Lambda usage. Or maybe controlling this is the core Vercel business model and it will never happen.

Any comments from someone knowledgeable would be greatly appreciated!

**Reply by zsiegel92** [reply]
I guess the answer is to contribute [Community runtimes](https://vercel.com/docs/functions/serverless-functions/runtimes#community-runtimes) using the [Runtime Developer Reference](https://github.com/vercel/vercel/blob/main/DEVELOPING_A_RUNTIME.md).

I guess unlike PHP, Rust, Deno, and Bash (!), Python's community has not to-date been enterprising enough to contribute any community runtimes!

**Reply by zsiegel92** [reply]
But per the quote in my previous comment, I guess this doesn't apply to serverless functions.

Again, any comment from Vercel about paths forward would be great.

**Comment by draew6** [comment]
Support modern python please. 3.9 feels ancient compared to 3.10 syntax.

**Comment by TheRobBrennan** [comment]
Any progress on this front? 3.9 is long in the tooth - would love to see 3.11.1 or newer

**Reply by XChikuX** [reply]
Fine. But, only 3.12 from 2024. 📦

**Reply by u8621011** [reply]
@XChikuX You mean vercel will support v3.12 from 2024?

**Comment by Flecart** [comment]
Updates?

**Comment by alex-goswag** [comment]
When will we have access to Python 3.11 and Python 3.12 runtimes on Serverless Functions?

**Reply by XChikuX** [reply]
I don't think they care bro.

**Reply by Flecart** [reply]
They should care....

**Reply by baseplate-admin** [reply]
They dont care because `vercel` is not known for it's `python` support. They offer it as a commodity service, i doubt anyone is using this feature at scale.

**Reply by dalist1** [reply]
They're cashing in on AI for now, give them a break.

**Comment by swarnava** [comment]
We are preparing for Python 3.12 and Ruby 3.3.0 soon. Stay tuned! 😄

**Reply by draew6** [reply]
Nice

**Reply by u8621011** [reply]
I cannot wait !~

**Comment by amyegan** [comment, marked as answer]
Our [Python Runtime](https://vercel.com/docs/functions/runtimes/python#python-version) now supports Python 3.12 🙌 

https://vercel.com/docs/deployments/build-image#runtime-support

**Reply by TheRobBrennan** [reply]
Eeee this is great 🎉 Thanks @amyegan

**Reply by XChikuX** [reply]
Thank you!

**Reply by ernests** [reply]
I don't know where to post this and will add it here.

Trying to deploy Python 3.12. FastAPI and in logs encountering this error:

```python
TypeError: Queue.__init__() got an unexpected keyword argument 'loop'
Traceback (most recent call last):
  File "/var/task/vc__handler__python.py", line 305, in vc_handler
    response = asgi_cycle(__vc_module.app, body)
  File "/var/task/vc__handler__python.py", line 202, in __call__
    self.app_queue = asyncio.Queue(loop=loop)
 ```
 
 It has been mentioned before by @adamstoffel  [above](https://github.com/orgs/vercel/discussions/639#discussioncomment-8238278)

**Reply by ernests** [reply]
This is the problem part. - https://github.com/vercel/vercel/blob/dc974b6797de0b6e90373c92e1f2bbdafcfc6687/packages/python/vc_init.py#L187

**Reply by amyegan** [reply]
Thanks for sharing a link to the code! I see `loop = asyncio.new_event_loop()` there, which makes me think it's related to this Stack Overflow post: https://stackoverflow.com/questions/60312374/what-are-all-these-deprecated-loop-parameters-in-asyncio

Do you have a [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example) that I can share with the team?

**Reply by ernests** [reply]
Thank you @amyegan for quick response. 
I could share ChatGPT refactored version :) but probably it would be better for some team member to have a look at this issue in wider context. 
Essentially, Python has optimised asyncio and deprecated usage of loops as you already pointed out with Stackoverflow issue. Vercel is not alone with this issue as other libraries are also updating code base to handle asyncio.

**Reply by amyegan** [reply]
I wasn't able to see the error in my original test app, but just tried with a fresh copy of the [Next.js FastAPI Starter](https://vercel.com/templates/next.js/nextjs-fastapi-starter) template and I can see it now. Sent your report and repro steps to the team. Thanks again for sharing it!

**Reply by winglian** [reply]
@amyegan Is there a GH issue we can follow for this?

**Reply by winglian** [reply]
here's a PR to get the ball rolling on this https://github.com/vercel/vercel/pull/11541
